File: week5/python/CourseSchedule_207.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        mapping = dict()
        visited = dict()
        for [u,v] in prerequisites:
            if u not in mapping:
                mapping[u] = list()
                
            mapping[u].append(v)
            visited[u] = 0
            visited[v] = 0
        
        def dfs(cur: int) -> bool:
            
            if visited[cur] == -1:
                return False
        
            if cur not in mapping:
                visited[cur] = 1
            
            if visited[cur] == 1:
                return True
        
            visited[cur] = -1
            for child in mapping[cur]:
                if not dfs(child):
                    return False
            
            visited[cur] = 1
            return True
        
        for course in mapping.keys():
            if not dfs(course):
                return False
        
        return True

File: week5/java/CourseSchedule_207.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if (prerequisites == null || prerequisites.length < 2) 
            return true;

        List[] map = new ArrayList[numCourses];
        for (int[] courses : prerequisites) {
            if (map[courses[0]] == null)
                map[courses[0]] = new ArrayList<Integer>();
            map[courses[0]].add(courses[1]);
        }

        int[] groupAssigned = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
            if (groupAssigned[i] == 0 && !canFinish(map, i, groupAssigned))
                return false;
        }
        return true;
    }
    
    private boolean canFinish(List<Integer>[] map, int index, int[] groupAssigned) {
        if (groupAssigned[index] != 0)   
            return groupAssigned[index] == 1;
        
        groupAssigned[index] = -1;
        if (map[index] != null) {
            for (Integer course : map[index]) {
                if (!canFinish(map, course, groupAssigned))
                    return false;
            }
        }
        groupAssigned[index] = 1;
        return true;
    }
    
}
